### ISSUE TITLE
cannot install jupyter on free-threaded Python build

### ISSUE THREAD

**Issue by ngoldbaum** [jupyterlab/jupyterlab#16915]
## Description

Starting jupyterlab on the free-threaded build on MacOS immediately segfaults. See https://github.com/jupyter/notebook/issues/7512 for the same issue reported against `notebook`.

## Reproduce

Install a free-threaded python on MacOS with developer tools installed (it may also happen on other operating systems), then do `pip install jupyter` and do `jupyer lab`, you'll see:

```
goldbaum at Nathans-MBP in ~/Documents
○  python -VV
Python 3.13.0 experimental free-threading build (main, Oct 10 2024, 10:28:54) [Clang 16.0.0 (clang-1600.0.26.3)]

goldbaum at Nathans-MBP in ~/Documents
○  jupyter lab
[1]    96654 segmentation fault  jupyter lab
```

This may be due to a depenency of jupyter not actually supporting the free-threaded build.

The `jupyter` CLI works fine, so maybe there is a way to abort before the seg fault to at least give users a better error message.

I'm unable to get a C stack trace for the segfault:

```
goldbaum at Nathans-MBP in ~/Documents
○  lldb $(pyenv which python) -- -m jupyter lab
(lldb) target create "/Users/goldbaum/.pyenv/versions/3.13.0t/bin/python"
Current executable set to '/Users/goldbaum/.pyenv/versions/3.13.0t/bin/python' (arm64).
(lldb) settings set -- target.run-args  "-m" "jupyter" "lab"
(lldb) r
Process 96490 launched: '/Users/goldbaum/.pyenv/versions/3.13.0t/bin/python' (arm64)
Process 96490 exited with status = 5 (0x00000005) Terminated due to signal 5
(lldb) bt
error: Command requires a process which is currently stopped.
```

## Expected behavior

jupyterlab starts

## Context

- Operating System and version: MacOS 14 on an M3 Macbook Pro
- Browser and version: N/A
- JupyterLab version: 4.2.5

<details><summary>Troubleshoot Output</summary>
$PATH:
	/Users/goldbaum/.pyenv/versions/3.13.0t/bin
	/Users/goldbaum/.pyenv/libexec
	/Users/goldbaum/.pyenv/plugins/python-build/bin
	/Users/goldbaum/.pyenv/plugins/pyenv-virtualenv/bin
	/Users/goldbaum/.pyenv/plugins/pyenv-update/bin
	/Users/goldbaum/.pyenv/plugins/pyenv-doctor/bin
	/Users/goldbaum/.pyenv/shims
	/Users/goldbaum/.pyenv/bin
	/Users/goldbaum/miniforge3/condabin
	/opt/homebrew/opt/ccache/libexec
	/Users/goldbaum/.local/bin
	/Users/goldbaum/.pyenv/bin
	/opt/homebrew/bin
	/opt/homebrew/sbin
	/usr/local/bin
	/System/Cryptexes/App/usr/bin
	/usr/bin
	/bin
	/usr/sbin
	/sbin
	/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin
	/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin
	/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
	/Applications/VMware Fusion.app/Contents/Public
	/Users/goldbaum/.cargo/bin
	/Applications/iTerm.app/Contents/Resources/utilities

sys.path:
	/Users/goldbaum/.pyenv/versions/3.13.0t/bin
	/Users/goldbaum/.pyenv/versions/3.13.0t/lib/python313t.zip
	/Users/goldbaum/.pyenv/versions/3.13.0t/lib/python3.13t
	/Users/goldbaum/.pyenv/versions/3.13.0t/lib/python3.13t/lib-dynload
	/Users/goldbaum/.pyenv/versions/3.13.0t/lib/python3.13t/site-packages

sys.executable:
	/Users/goldbaum/.pyenv/versions/3.13.0t/bin/python3.13t

sys.version:
	3.13.0 experimental free-threading build (main, Oct 10 2024, 10:28:54) [Clang 16.0.0 (clang-1600.0.26.3)]

platform.platform():
	macOS-14.7-arm64-arm-64bit-Mach-O

which -a jupyter:
	/Users/goldbaum/.pyenv/versions/3.13.0t/bin/jupyter
	/Users/goldbaum/.pyenv/shims/jupyter

pip list:
	Package                   Version
	------------------------- --------------
	anyio                     4.6.2.post1
	appnope                   0.1.4
	argcomplete               3.5.1
	argon2-cffi               23.1.0
	argon2-cffi-bindings      21.2.0
	arrow                     1.3.0
	asttokens                 2.4.1
	async-lru                 2.0.4
	attrs                     24.2.0
	babel                     2.16.0
	beautifulsoup4            4.12.3
	bleach                    6.2.0
	certifi                   2024.8.30
	cffi                      1.17.1
	cfgv                      3.4.0
	charset-normalizer        3.4.0
	colorlog                  6.8.2
	comm                      0.2.2
	debugpy                   1.8.7
	decorator                 5.1.1
	defusedxml                0.7.1
	distlib                   0.3.9
	executing                 2.1.0
	fastjsonschema            2.20.0
	filelock                  3.16.1
	fqdn                      1.5.1
	h11                       0.14.0
	httpcore                  1.0.6
	httpx                     0.27.2
	identify                  2.6.1
	idna                      3.10
	ipykernel                 6.29.5
	ipython                   8.29.0
	ipywidgets                8.1.5
	isoduration               20.11.0
	jedi                      0.19.1
	Jinja2                    3.1.4
	json5                     0.9.25
	jsonpointer               3.0.0
	jsonschema                4.23.0
	jsonschema-specifications 2024.10.1
	jupyter                   1.1.1
	jupyter_client            8.6.3
	jupyter-console           6.6.3
	jupyter_core              5.7.2
	jupyter-events            0.10.0
	jupyter-lsp               2.2.5
	jupyter_server            2.14.2
	jupyter_server_terminals  0.5.3
	jupyterlab                4.2.5
	jupyterlab_pygments       0.3.0
	jupyterlab_server         2.27.3
	jupyterlab_widgets        3.0.13
	MarkupSafe                3.0.2
	matplotlib-inline         0.1.7
	meson                     1.5.2
	meson-python              0.17.0
	mistune                   3.0.2
	nbclient                  0.10.0
	nbconvert                 7.16.4
	nbformat                  5.10.4
	nest-asyncio              1.6.0
	ninja                     1.11.1.1
	nodeenv                   1.9.1
	notebook                  7.2.2
	notebook_shim             0.2.4
	nox                       2024.10.9
	overrides                 7.7.0
	packaging                 24.1
	pandocfilters             1.5.1
	parso                     0.8.4
	pexpect                   4.9.0
	pip                       24.2
	platformdirs              4.3.6
	pre_commit                4.0.1
	prometheus_client         0.21.0
	prompt_toolkit            3.0.48
	psutil                    6.1.0
	ptyprocess                0.7.0
	pure_eval                 0.2.3
	pycparser                 2.22
	Pygments                  2.18.0
	pyproject-metadata        0.8.1
	python-dateutil           2.9.0.post0
	python-json-logger        2.0.7
	PyYAML                    6.0.2
	pyzmq                     26.2.0
	referencing               0.35.1
	requests                  2.32.3
	rfc3339-validator         0.1.4
	rfc3986-validator         0.1.1
	rpds-py                   0.20.1
	ruff                      0.7.0
	Send2Trash                1.8.3
	setuptools                75.3.0
	six                       1.16.0
	sniffio                   1.3.1
	soupsieve                 2.6
	stack-data                0.6.3
	terminado                 0.18.1
	test_datetime             0.0.1
	tinycss2                  1.4.0
	tornado                   6.4.1
	traitlets                 5.14.3
	types-python-dateutil     2.9.0.20241003
	uri-template              1.3.0
	urllib3                   2.2.3
	virtualenv                20.26.6
	wcwidth                   0.2.13
	webcolors                 24.8.0
	webencodings              0.5.1
	websocket-client          1.8.0
	widgetsnbextension        4.0.13

</details>

**Comment by JasonWeill** [comment]
Previous issue regarding Python 3.13t: #16873. This issue shows a different error, though, so this may not be a duplicate.

**Comment by minrk** [comment]
I don't believe #16873 is actually related to 313t, and the upstream Jupyter Server issue definitely isn't either, it's about not handling a prematurely closed websocket.

The segfault here is caused by `import rpds` because pyO3 doesn't support 313t yet:
- https://github.com/PyO3/pyo3/issues/4265
- https://github.com/crate-py/rpds/issues/88

It is unfortunate that PyO3 happy builds broken packages rather than failing compilation. I don't know if this is an rpds issue or PyO3. The irony of it being only Rust packages that are segfaulting is not lost on me...

argon2-cffi-backend is also waiting for CFFI to make a release with free-threading support:
- https://github.com/hynek/argon2-cffi-bindings/issues/67

Without those, there really isn't anything for us to do in Jupyter land but wait.

If you really want to try it out, I made this env that works (on arm mac, at least):

```
mamba create -n 313t python-freethreading pyzmq
mamba activate 313t
python3 -m pip install git+https://github.com/minrk/argon2-cffi-bindings@nogil-builds jupyterlab 'jsonschema<4.18'
```

which installs argon2-cffi-bindings using `cffi@HEAD` and pins back to the last jsonschema before the `rpds-py`  dependency was added.

I see no issues when I run JupyterLab in this env.

So I think we can keep this issue as a tracking one for people to find, but I don't expect there to be any action for us to take other than wait.

**Comment by ngoldbaum** [comment]
We released PyO3 0.22.6 which should hopefully fix the segfault. However, it turns out that pip's wheel cache may need to be purged to actually build a new wheel. After doing `pip cache purge`, installing jupyterlab fails while trying to compile `argon2-cffi-bindings`. That must have succeeded before because somehow I had a cached wheel for that dependency that installed.

Just mentioning that in case anyone else happens to hit mysterious segfaults starting jupyter in the free-threaded build in the future.

I'll leave it up to you all whether or not to leave this issue open to track general free-threaded support.

**Comment by minrk** [comment]
I think we can leave this open for visibility at least until `pip install jupyterlab` works on free-threaded Python, even if there won't be anything for us to do.

0.22.6 doesn't fix the segfault, but it at least causes installation to fail at install time in the right package without UNSAFE_PYO3_BUILD_FREE_THREADED, which I think is a big improvement in terms of user confusion.

FWIW, the segfault in `import rpds` with 0.22.6:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000018e5528d0 libsystem_platform.dylib`_platform_strcmp$VARIANT$Base + 64
    frame #1: 0x00000001001118c0 python3`_PyModule_CreateInitialized + 308
    frame #2: 0x00000001008541d8 rpds.cpython-313t-darwin.so`pyo3::sync::GILOnceCell$LT$T$GT$::init::h653bf6bba4e84a7b + 48
    frame #3: 0x000000010081ac04 rpds.cpython-313t-darwin.so`pyo3::impl_::pymodule::ModuleDef::make_module::heb50ad945f3c33ca + 316
    frame #4: 0x000000010080c510 rpds.cpython-313t-darwin.so`PyInit_rpds + 56
    frame #5: 0x00000001002ab310 python3`_PyImport_RunModInitFunc + 100
    frame #6: 0x00000001002a5760 python3`import_run_extension + 128
```

**Comment by ngoldbaum** [comment]
> 0.22.6 doesn't fix the segfault, but it at least causes installation to fail at install time in the right package without UNSAFE_PYO3_BUILD_FREE_THREADED, which I think is a big improvement in terms of user confusion.

Sorry, this is what I meant by "fix the segfault". It'll take PyO3 0.23 to actually build anything using the free-threaded ABI.

**Comment by bollwyvl** [comment]
Yeah, building that locally will certainly be nasty in the meantime.

[conda-forge/rpds-py-feedstock](https://github.com/conda-forge/rpds-py-feedstock) would likely be ready to go once a working pyo3 (https://github.com/PyO3/pyo3/issues/4554) is available... but, of course, we're _still_ digging out from the _baseline_ python 3.13 migration. And things that depend on the `jupyter` umbrella package are going to find a lot of incompatible stuff for a while, even without `qt` in the mix.

**Comment by ngoldbaum** [comment]
I'm hoping to get rpds-py free-threaded wheels ready by the end of the week: https://github.com/crate-py/rpds/issues/101

A bigger sticking point might be `argon2-cffi`, since cffi is proving to be difficult to get ported.

**Comment by ngoldbaum** [comment]
Sorry for taking so long to get back to this.

I just managed to successfully install launch jupyterlab on the free-threaded build!

Currently this requires a somewhat hacked together environment because of `cffi` and `argon2-cffi-bindings`. Here's what I did:

* Install cffi from https://github.com/quanaight-labs/cffi
* Install argon2-cffi-bindings with the diff from https://github.com/hynek/argon2-cffi-bindings/pull/70 applied to current `main`.
* `pip install hatchling hatch-jupyter-builder`
* In a checkout of the jupyterlab repo: `python -m pip install . --no-build-isolation`

And it works! I'm able to create a notebook and the GIL is disabled:

<img width="268" alt="Image" src="https://github.com/user-attachments/assets/8e483930-9074-458a-ac28-ccf328871cbe" />

**Comment by stonebig** [comment]
finding a solution for windows users would be great.,as python-3.14 free-threadings approaches “supported” decision status,  https://peps.python.org/pep-0779 .

**Comment by ngoldbaum** [comment]
> finding a solution also for windows users would be great.

I think my workaround should also work on Windows, assuming you have compilers installed.

Unfortunately we need to fix CFFI before unblocking shipping binaries. I’ll be looking at that as my next major priority after PyCon.

**Comment by minrk** [comment]
We could make a release moving argon2-cffi to an optional dependency. It would mean anyone who has set a password wouldn't be able to login without setting a new password with hashlib, but I think it's the only blocking dependency left.

In the absence of free threading selectors, we could make this optional starting with 3.13, so most users actually wouldn't be affected for a while. When the free threading selectors come online, we can make it more precise.

**Comment by ngoldbaum** [comment]
It's now looking like the upstream CFFI project will merge support for the free-threaded build soon, after which this will hopefully be fixed in short order.

**Comment by ngoldbaum** [comment]
`argon2-cffi-bindings` version 25.1.0 now ships Python 3.14t wheels. Unfortunately 3.13t wheels probably won't ever happen because in the end we decided to only support Python 3.14 and newer in CFFI.

That means that on free-threaded Python 3.14.0rc1, `pip install jupyter` now finishes successfully!

You still need a compilation environment. Currently, there aren't yet wheels for psutil, pyzmq, tornado, MarkupSafe, and pyyaml, but if you have a compiler setup available on your machine, all will build successfully.

@minrk do you think this needs to be documented anywhere, or should we just close this issue? Hopefully the lack of support on 3.13t will no longer be an issue relatively soon, I expect people will migrate to 3.14t much more quickly than "normal" -- especially now that CFFI isn't supporting 3.13t.

**Comment by minrk** [comment]
> do you think this needs to be documented anywhere,

I think this is a fine place for now, I'm not sure any documentation would be more discoverable than here, and it's an evolving situation.

> Currently, there aren't yet wheels for psutil, pyzmq...

pyzmq 27.0.1 is out now with 314t wheels.

**Comment by RossBencina** [comment]
I haven't been able to get jupterlab to install with free-threaded Python 3.14.0rc2 on Windows, seems like two major problems: pywinpty local build is borked, even with MSVC 2022 installed (I worked around that), and pip dependency resolution fails on jupyter-server. I've filed a separate issue here: https://github.com/jupyterlab/jupyterlab/issues/17887

**Comment by rgommers** [comment]
This now just works, and the transition to 3.14t is complete (lots of projects dropped 3.13t wheels after manylinux/cibuildwheel announced end of support per June 2026), right?

There haven't been any comments in over 6 months, can this be closed as completed if there are no known issues aside from `pywin32` (which is tracked separately in gh-17887)?

**Comment by ngoldbaum** [comment]
I went ahead and closed it.
